mqtt/update_agent_client.go

```diff
@@ -0,0 +1,229 @@
+// Copyright (c) 2023 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package mqtt
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/eclipse-kanto/update-manager/api"
+	"github.com/eclipse-kanto/update-manager/api/types"
+	"github.com/eclipse-kanto/update-manager/logger"
+
+	pahomqtt "github.com/eclipse/paho.mqtt.golang"
+	"github.com/google/uuid"
+)
+
+const prefixInitCurrentStateID = "initial-current-state-"
+
+const (
+	suffixDesiredState         = "/desiredstate"
+	suffixDesiredStateCommand  = "/desiredstate/command"
+	suffixCurrentState         = "/currentstate"
+	suffixCurrentStateGet      = "/currentstate/get"
+	suffixDesiredStateFeedback = "/desiredstatefeedback"
+
+	disconnectQuiesce uint = 10000
+)
+
+type mqttClient struct {
+	mqttPrefix string
+	mqttConfig *ConnectionConfig
+	pahoClient pahomqtt.Client
+}
+
+type updateAgentClient struct {
+	*mqttClient
+	domain  string
+	handler api.UpdateAgentHandler
+}
+
+// NewUpdateAgentClient instantiates a new UpdateAgentClient instance using the provided configuration options.
+func NewUpdateAgentClient(domain string, config *ConnectionConfig) api.UpdateAgentClient {
+	client := &updateAgentClient{
+		mqttClient: &mqttClient{
+			mqttPrefix: domainAsTopic(domain),
+			mqttConfig: config,
+		},
+		domain: domain,
+	}
+	client.pahoClient = newClient(config, client.onConnect)
+	return client
+}
+
+func (client *updateAgentClient) topic(topicSuffix string) string {
```

Comment:
I guess generating the topics once at the creation of the client will be more efficient than calculating them every time. What do you think?